src/plugins/stdioproc/README.md

```diff
@@ -0,0 +1,201 @@
+It must also contain `exports/_` keys for every operation that is implemented by the child.
+All keys in `[contract]` should be below `system:/elektra/modules/stdioproc`.
+The parent will rename these keys appropriately.
+
+A child that implements all operations should include these keys (in addition to the relevant `infos` keys):
+
+```
+system:/elektra/modules/stdioproc/exports/open = 1
+system:/elektra/modules/stdioproc/exports/close = 1
+system:/elektra/modules/stdioproc/exports/get = 1
+system:/elektra/modules/stdioproc/exports/set = 1
+```
+
+After this initial handshake, the child should simply wait for further requests from the parent.
+
+> **Note**: Under normal circumstances this handshake will always be followed by an `open` request immediately.
+
+### Operation
+
+When the parent needs the child to process an operation (open, get, ...), it will send a request like this:
+
+```
+Parent > Child
+
+(opname)
+[parent]
+[data]
+```
+
+Here `(opname)` is one of `open`, `get`, `set` or `close` and describes the operation that shall be performed by the child.
+The keyset `[parent]` always consists of a single key, namely the `parentKey` (or `errorKey`) that was passed to the plugin.
+Finally, `[data]` is the keyset that was passed to the plugin.
+The `[data]` keyset is not present in `open` and `close` operations, since those don't receive a `KeySet` in the C API.
+However, in the `open` operation `[data]` is replaced `[config]` which is the `KeySet` returned by `elektraPluginGetConfig` in the C API.
+This is needed, because the child process cannot request the config keyset otherwise.
+
+The child should then perform the requested operation and respond with
+
+```
+Child > Parent
+
+(result)
+[parent]
+[data]
+```
+
+Here `(result)` is one of `success`, `noupdate` and `error`, which correspond to `ELEKTRA_PLUGIN_STATUS_SUCCESS`, `ELEKTRA_PLUGIN_STATUS_NO_UPDATE` and `ELEKTRA_PLUGIN_STATUS_ERROR` respectively.
+The keysets `[parent]` and `[data]` are the modified versions of the one sent by the parent.
+
+### Termination
+
+When the parent no longer needs the child process, it will send a final termination request.
+
+```
+Parent > Child
+
+ELEKTRA_STDIOPROC TERMINATE
+```
+
+The child process should now exit (and thereby close its ends of the `stdin`/`stdout` pipes).
+
+> **Note:** Under normal circumstances this only happens, when plugin is being closed.
+
+### Errors
+
+If an unexpected error occurs on either side of the protocol, the connection should be terminated and the child process shall exit.
+
+## Examples
+
+```sh
+# mount the Whitelist Java Plugin via stdioproc
+sudo kdb mount config.file user:/tests/stdioproc dump stdioproc 'app=/usr/bin/java' 'args=#3' 'args/#0=-cp' "args/#1=$BUILD_DIR/src/bindings/jna/plugins/whitelist/build/libs/whitelist-$(kdb --version | sed -nE 's/KDB_VERSION: (.+)/\1/gp')-all.jar" 'args/#2=org.libelektra.stdioproc.StdIoProcApp' 'args/#3=org.libelektra.plugin.WhitelistPlugin'
```

Comment:
Can you please write a `mount-java` wrapper that takes the ugliness out of this?

Reply:
In theory we could do something like

```sh
sudo kdb mount-java config.file user:/tests/stdioproc kdb:dump java:org.libelektra.plugin.WhitelistPlugin 
```

The `java:` and `kdb:` prefixes are to differentiate between `Java` plugins and others. This would also only work for Java plugins that are part of this repo. If you want to use plugins from other JARs, you'd have to define the `CLASSPATH` environment variable accordingly.

However, I don't think this example should be changed at all (adding `mount-java` is still a good idea). This example is not supposed to show how Java plugins are mounted. It is supposed to be an example of how to use `stdioproc`. Therefore, it should not hide the actual `stdioproc` config in some helper script.

Reply:
@kodebach wrote:
> In theory we could do something like
> 
> sudo kdb mount-java config.file user:/tests/stdioproc kdb:dump java:org.libelektra.plugin.WhitelistPlugin

I think we can assume `kdb:` if no `java:` is given?

> The |java:| and |kdb:| prefixes are to differentiate between |Java| 
> plugins and others. This would also only work for Java plugins that are 
> part of this repo. If you want to use plugins from other JARs, you'd 
> have to define the |CLASSPATH| environment variable accordingly.
> 
> However, I don't think this example should be changed at all (adding 
> |mount-java| is still a good idea). This example is not supposed to show 
> how Java plugins are mounted. It is supposed to be an example of how to 
> use |stdioproc|. Therefore, it should not hide the actual |stdioproc| 
> config in some helper script.

I agree, in the process plugin there should be the full example without any helper scripts. Could you please add/extend a Java-Plugin Tutorial where you describe `mount-java` and also how to set CLASSPATH for out-of-source-tree Java Plugins?

Reply:
> I think we can assume kdb: if no java: is given?

I don't think there is technical reason why `java:com.example.Foo` couldn't be the name of a plugin written in C. The `.so` file certainly could contain `.` and `:`. But we could make `kdb:` optional. Then we'd have:
- `kdb:java:com.example.Foo` means C plugin named `java:com.example.Foo`
- `java:com.example.Foo` means Java plugin implemented in the class `com.example.Foo`
- `foo` and `kdb:foo` both mean C plugin named `foo`


>  Could you please add/extend a Java-Plugin Tutorial where you describe mount-java and also how to set CLASSPATH for out-of-source-tree Java Plugins?

Will do